### DISCUSSION TITLE
How to define example_input_array for computational graph

### DISCUSSION THREAD

**Discussion by minwang-ai** [Lightning-AI/pytorch-lightning#10405] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Could you please give me an example for defining self.example_input_array according to the document?

When and where should I log computational graph, train/Val/test step?

log_graph (bool) – Adds the computational graph to tensorboard. This requires that the user has defined the self.example_input_array attribute in their model.

**Comment by rohitgr7** [comment]
you can simply assign it to a variable:
```py
class LitModel(LightningModule):
    def __init__(self, ...):
        ...
        self.example_input_array = ...
```

**Reply by minwang-ai** [reply]
Do you mean define a random tensor with same size as input?

Apart from that, I only need to set the log_graph flag to True, right?
Could you tell me when and where I will log computational graph, train/Val/test step?

**Reply by minwang-ai** [reply]
Should I specify log in test_step or train_step?

def test_step(self, batch, batch_idx):
    ...
    self.logger.log_graph(self)
    ...

**Reply by minwang-ai** [reply]
> Do you mean define a random tensor with same size as input?
What if I pass a dictionary to my model as input and use different values for CNN input and pointNet input and fusing two network’s feature map together?

**Reply by rohitgr7** [reply]
> Do you mean define a random tensor with same size as input?

yes

> Apart from that, I only need to set the log_graph flag to True, right?

yes

> Should I specify log in test_step or train_step?

no need, it will happen automatically for you.

> What if I pass a dictionary to my model as input and use different values for CNN input and pointNet input and fusing two network’s feature map together?

I think it depends on your forward method. If it can handle dict input and it should work I guess.

**Reply by minwang-ai** [reply]
Thank you for your reply!



> > Do you mean define a random tensor with same size as input?
> 
> yes
> 
> > Apart from that, I only need to set the log_graph flag to True, right?
> 
> yes
> 
> > Should I specify log in test_step or train_step?
> 
> no need, it will happen automatically for you.
> 
> > What if I pass a dictionary to my model as input and use different values for CNN input and pointNet input and fusing two network’s feature map together?
> 
> I think it depends on your forward method. If it can handle dict input and it should work I guess.

My forward method can handle dict input but I don’t know how to define self.example_input_array in this case, since my input is a nested dictionary, each value is also a dictionary and each value is a tensor.

**Reply by rohitgr7** [reply]
```py
class LitModel(LightningModule):
    def __init__(self, ...):
        ...
        self.example_input_array = {'key1': {'nested_key1': torch.Tensor(...)}, 'key2': ...}
```
this doesn't work?

**Comment by sascharo** [comment]
> ```python
> ```python
> class LitModel(LightningModule):
>     def __init__(self, ...):
>         ...
>         self.example_input_array = {'key1': {'nested_key1': torch.Tensor(...)}, 'key2': ...}
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> this doesn't work?
> ```
Not for me.
